python/treelite/gtil/gtil.py

```diff
@@ -26,34 +60,74 @@ def predict(model: Model, data: np.ndarray, nthread: int = -1, pred_margin: bool
         2D NumPy array, with which to run prediction
     nthread : :py:class:`int <python:int>`, optional
         Number of CPU cores to use in prediction. If <= 0, use all CPU cores.
-    pred_margin : :py:class:`bool <python:bool>`, optional
-        Whether to produce raw margin scores
+    pred_margin : bool, optional
+        Deprecated. Set predict_type="raw" instead.
+    predict_type : string, optional
+        One of the following:
+        * "default": Usual prediction method. Sum over trees and apply post-processing.
+        * "raw": Sum over trees, but don't apply post-processing; get raw margin scores
+                 instead.
+        * "leaf_id": Output one (integer) leaf ID per tree.
+        * "score_per_tree": Output one or more margin scores per tree.
 
     Returns
     -------
     prediction : :py:class:`numpy.ndarray` array
         Prediction
     """
-    assert isinstance(model, Model)
-    assert isinstance(data, np.ndarray)
-    assert len(data.shape) == 2
-    data = np.array(data, copy=False, dtype=np.float32, order='C')
+    # Parameter validation
+    if pred_margin is not None:
+        if predict_type:
+            raise ValueError(
+                "Cannot specify pred_margin and predict_type at the same "
+                "time. Please set predict_type only."
+            )
+        warnings.warn(
+            "The pred_margin argument is deprecated. Please use "
+            'predict_type="raw" instead.',
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        predict_type = "raw"
+    if not predict_type:
+        predict_type = "default"
```

Comment:
I would check explicitly against None to not confuse with falsy values such as _empty string_ or `False`.
Further, I would suggest to capture all valid predict_types in a [`StrEnum`](https://docs.python.org/3/library/enum.html#enum.StrEnum).

```suggestion
    if predict_type is None:
        predict_type = "default"
    elif predict_type not in list(_PREDICT_TYPES):
        raise ValueError(f"Unknown predict_type: {predict_type}.")
```

Reply:
No longer relevant, as I'm creating separate functions `predict_leaf` and `predict_per_tree`.